File: 백준/Gold/2580. 스도쿠/스도쿠.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.StringTokenizer;
import java.util.stream.Collectors;

public class Main {

	static int[][] board = new int[9][9];

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		for (int i = 0; i < 9; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			for (int j = 0; j < 9; j++) {
				board[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		sudoku(0, 0);
	}

	private static boolean sudoku(int row, int col) {

		if (col == 9) {
			return sudoku(row + 1, 0);
		}

		if (row == 9) {
			printBoard();
			return true;
		}

		if (board[row][col] == 0) {
			for (int num = 1; num <= 9; num++) {
				if (isPossible(row, col, num)) {
					board[row][col] = num;
					if (sudoku(row, col + 1)) {
						return true;
					}
					board[row][col] = 0;
				}
			}
			return false;
		} else {
			return sudoku(row, col + 1);
		}
	}

	private static boolean isPossible(int row, int col, int num) {

		int rowStart = (row / 3) * 3;
		int colStart = (col / 3) * 3;

		for (int i = 0; i < 9; i++) {
			if (board[row][i] == num || board[i][col] == num || board[rowStart + (i / 3)][colStart + (i % 3)] == num) {
				return false;
			}
		}

		return true;
	}

	private static void printBoard() {
		StringBuilder sb = new StringBuilder();
		for (int i = 0; i < 9; i++) {
			for (int j = 0; j < 9; j++) {
				sb.append(board[i][j]).append(" ");
			}
			sb.append("\n");
		}
		System.out.println(sb.toString());
	}
}

File: 백준/Gold/2580. 스도쿠/스도쿠.py
import sys

lines = sys.stdin.readlines()
board = [list(map(int, line.rstrip().split())) for line in lines]


def is_possible(row, col, num):
    row_start = (row // 3) * 3
    col_start = (col // 3) * 3

    for i in range(9):
        if (
            board[row][i] == num
            or board[i][col] == num
            or board[row_start + (i // 3)][col_start + (i % 3)] == num
        ):
            return False
    return True


def sudoku(row, col):
    if col == 9:
        return sudoku(row + 1, 0)

    if row == 9:
        # [print(*board[i], sep=' ') for i in range(9)]
        output = '\n'.join([' '.join(map(str, row)) for row in board])
        print(output)
        return True

    if board[row][col] == 0:
        for num in range(1, 10):
            if is_possible(row, col, num):
                board[row][col] = num
                if sudoku(row, col + 1):
                    return True
                board[row][col] = 0
        return False
    else:
        return sudoku(row, col + 1)


sudoku(0, 0)
